### PR TITLE
Votes ddos protection

### PR DESCRIPTION
**Pull request by JakubFornadel**
## Purpose

This PR is removing concept of unverified votes. Previously we would accept every vote and without any validation we would put it into the unverified queue, which could lead to memory grow...

Many possible ddos attacks are now eliminated by these changes, but not all of them. Anyway we could probably merge this PR so Dmitry can use it and then we can have separate PR with additional changes...

## How does the solution address the problem

In the new version:

- each vote contains also pbft period so we can validate it against dpos contract as soon as we receive it
- each vote is fully validated in packet handlers and if validation fails, vote is dropped
- there is no unverified queue for votes anymore
- each voter can create only 1 unique vote per round & step & voter address. Exception is next_vote - we allow 2 unique votes - 1 for some specific block hash and 1 for NULL_BLOCK_HASH
- if some voter already sent standard vote with round N, any new vote from him with round N-1 is dropped (reward & sync next) votes are exceptions

##### Standard vote:
`vote_round >= current_pbft_round`
// Here should be vote_period > current_pbft_period but it breaks some tests due to missing rewards vote -> rewards votes need to be revisited
`vote_period >= current_pbft_period && vote_period <= current_pbft_period + dpos_delay(5)`

##### Reward vote:
`vote_round < current_pbft_round`
`vote_period == current_pbft_period`

##### Sync next vote:
`vote_round >= current_pbft_round - 1`
`vote_period >= current_pbft_period && vote_period <= current_pbft_period + dpos_delay(5)`

## Problems

- We differentiate standard & reward simply by checking round, `if vote_round < current_pbgt_round` we identify vote as reward vote and use validation for rewards votes -> this might be a problem as some out of sync node might send us standard vote and we process it as reward vote.... https://github.com/Taraxa-project/taraxa-node/issues/1880
- when running local-net I observed many errors on reward votes validation, when `vote_round < current_pbft_round` but `vote_period > current_pbft_period` -> this does not make sense: how can vote has older round than current_pbft_round and newer period than current_pbft_period ? 
- Standard & sync next vote -> as the rule is `vote_round >= current_pbft_round`, attacker might just keep increasing pbft round and these votes would pass validation and fill up verified queues. How to solve this ? We could put here at least some upper boundary, e.g.  `vote_round <= current_pbft_round + 1000` or some big number ?  

## Things to review properly:
I am now putting period into each placeVote_ call -> this period is then in the vote. In some votes like soft/cert votes I am using period from proposes pbft block, in other votes when we for example vote for NULL_BLOCK_HASH I am using current pbft period -> we should think about this approach if it is 100% valid !

### CODE REVIEW

**Review by justinsnapp — COMMENTED** [review]
@JakubFornadel  thanks for the excellent write-up on this PR.     I think everything you have done should definitely allow us to remove the queue for unvalidated votes.    But perhaps it still allows for a node to send us too many validatable votes, by sending us votes for wildly in the future rounds?   It seems a node can still can send votes from many rounds in the future, even if they get to speak only sometimes by churning through lots of rounds they could still amass a lot of votes to send?   

Unless we are out of sync they should be at most one round ahead of us.   So any node sending us votes for a round far ahead of local nodes round, the local node could request to sync to their round by getting 2t+1 next votes from previous round or else consider such a node malicious ... ie by getting a sync next vote packet.    Alteneratively since we send status packets (still), we could put our step and round in that and if the round was "wildly" (allowing for the status packet update interval) behind the local node we would send them a sync next votes packet and not relay votes way ahead.  And on the receiving side if we got votes for a round "wildly" ahead of our round we would mark that node as malicious.

You mentioned being a bit hazy as how to define the correct period for vote for each round to send.  I think the proper definition is that it **should be that it doesn't change during the round... so should be the same throughout the round, and be the period at the start of the round, before any possible growth in the chain by commit of a new pbft block**

**Comment by JakubFornadel** [comment]
> @JakubFornadel thanks for the excellent write-up on this PR. I think everything you have done should definitely allow us to remove the queue for unvalidated votes. But perhaps it still allows for a node to send us too many validatable votes, by sending us votes for wildly in the future rounds? It seems a node can still can send votes from many rounds in the future, even if they get to speak only sometimes by churning through lots of rounds they could still amass a lot of votes to send?
> 
> Unless we are out of sync they should be at most one round ahead of us. So any node sending us votes for a round far ahead of local nodes round, the local node could request to sync to their round by getting 2t+1 next votes from previous round or else consider such a node malicious ... ie by getting a sync next vote packet. Alteneratively since we send status packets (still), we could put our step and round in that and if the round was "wildly" (allowing for the status packet update interval) behind the local node we would send them a sync next votes packet and not relay votes way ahead. And on the receiving side if we got votes for a round "wildly" ahead of our round we would mark that node as malicious.
> 
> You mentioned being a bit hazy as how to define the correct period for vote for each round to send. I think the proper definition is that it **should be that it doesn't change during the round... so should be the same throughout the round, and be the period at the start of the round, before any possible growth in the chain by commit of a new pbft block**

-  Yes the problem with future rounds/steps is still not resolved but that could separate PR. This PR does not make the node more vulnerable to any new attack compared to previous version so even if we merge this it is going to be huge improvement & Dmitry can resolve conflicts while we work on fix for future rounds & steps. 
- I actually used this design before and the problem I had was that at the beginning of a new round I set period based on node pbft chain size, which is a problem because nodes might have different chain sizes... When I think about it now maybe we could set it based on periods from next votes. Currently we determine round based on 2t+1 next votes, what if we dont check just rounds, but also periods from these 2t+1 next votes and determine  round & period based on that ?

**Review by kstdl — APPROVED** [review]

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 1]
```diff
@@ -1110,14 +1123,50 @@ blk_hash_t PbftManager::generatePbftBlock(const blk_hash_t &prev_blk_hash, const
...
+      err << "No vrf key mapped for vote author " << vote->getVoterAddr();
+      return {false, err.str()};
+    }
+
+    // TODO[1896]: final_chain_->dpos... ret values should be cached as it is heavily used and most of the time it
+    // returns the same value!
+    const uint64_t voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(vote_period, vote->getVoterAddr());
+    const uint64_t total_dpos_votes_count = final_chain_->dpos_eligible_total_vote_count(vote_period);
+    const uint64_t pbft_sortition_threshold = getPbftSortitionThreshold(vote->getType(), vote_period);
+
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1150]: I think this should be called only when `vote_period != dpos_period_` as we currently have some kind of caching we should use it

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 2]
```diff
@@ -1129,17 +1178,56 @@ uint64_t PbftManager::getThreshold(PbftVoteTypes vote_type) const {
...
+  switch (vote_type) {
+    case propose_vote_type:
+      return std::min<uint64_t>(NUMBER_OF_PROPOSERS, total_dpos_votes_count);
+    case soft_vote_type:
+    case cert_vote_type:
+    case next_vote_type:
+    default:
+      return std::min<uint64_t>(COMMITTEE_SIZE, total_dpos_votes_count);
+  }
+}
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1193]: Having getCurrentPbftSortitionThreshold and getPbftSortitionThreshold does not make sense as it shares 90% of code logic, this could lead to errors where fix switch case just for one of the functions

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 3]
```diff
@@ -1129,17 +1178,56 @@ uint64_t PbftManager::getThreshold(PbftVoteTypes vote_type) const {
...
+  uint64_t total_dpos_votes_count = 0;
+  uint64_t pbft_sortition_threshold = 0;
+
+  try {
+    // TODO[1896]: final_chain_->dpos... ret values should be cached as it is heavily used and most of the time it
+    // returns the same value!
+    voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(period, node_addr_);
+    total_dpos_votes_count = final_chain_->dpos_eligible_total_vote_count(period);
+    pbft_sortition_threshold = getPbftSortitionThreshold(vote_type, period);
+
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1212]: again this could be check if it is not already pre-cached

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 4]
```diff
@@ -1398,17 +1485,18 @@ blk_hash_t PbftManager::identifyLeaderBlock_() {
       continue;
     }
 
-    leader_candidates.emplace_back(std::make_pair(getProposal(v), proposed_block_hash));
+    leader_candidates.emplace_back(std::make_pair(getProposal(v), v));
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1488]: why don't just use `getPeriod()` ? this should be more efficient than whole vote object

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 5]
```diff
@@ -1628,7 +1716,6 @@ void PbftManager::pushSyncedPbftBlocksIntoChain() {
       net->setSyncStatePeriod(period);
 
       if (executed_pbft_block_) {
-        vote_mgr_->removeVerifiedVotes();
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1631]: we do not need to clean those votes?

File: libraries/core_libs/network/src/network.cpp [hunk 6]
```diff
@@ -13,32 +13,33 @@
 
 namespace taraxa {
 
-Network::Network(NetworkConfig const &config, const h256 &genesis_hash,
+Network::Network(const FullNodeConfig &config, const h256 &genesis_hash,
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/network.cpp:16]: sorry I missed why did you change this? I did not see where it is used

File: tests/pbft_manager_test.cpp [hunk 7]
```diff
@@ -397,24 +383,8 @@ TEST_F(PbftManagerTest, check_get_eligible_vote_count) {
...
-      expected_eligible_total_vote++;
-    }
-    EXPECT_HAPPENS({120s, 1s}, [&](auto &ctx) {
-      for (auto &node : nodes) {
-        if (ctx.fail_if(node->getDB()->getNumTransactionExecuted() != trxs_count)) {
-          return;
-        }
-      }
-    });
-  }
```
Comment (MatusKysel) [inline tests/pbft_manager_test.cpp:417]: same here

File: libraries/types/vote/include/vote/vrf_sortition.hpp [hunk 8]
```diff
@@ -17,44 +17,35 @@ enum PbftVoteTypes : uint8_t { propose_vote_type = 0, soft_vote_type, cert_vote_
 /**
  * @brief VrfPbftMsg struct uses vote type, PBFT round, and PBFT step to generate a message for doing VRF sortition.
  */
-struct VrfPbftMsg {
+class VrfPbftMsg {
```
Comment (MatusKysel) [inline libraries/types/vote/include/vote/vrf_sortition.hpp:20]: if you changing it to class you should also rename member variables

File: libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp [hunk 9]
```diff
@@ -335,39 +271,35 @@ class VoteManager {
   using UpgradableLock = boost::upgrade_lock<boost::shared_mutex>;
   using UpgradeLock = boost::upgrade_to_unique_lock<boost::shared_mutex>;
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp:272]: you should remove those as we are no longer using them

File: libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp [hunk 10]
```diff
@@ -480,102 +403,34 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
...
-  if (!verifyRewardVote(vote)) {
-    return false;
-  }
-
-  {
-    std::unique_lock lock(reward_votes_mutex_);
-    reward_votes_.insert({vote_hash, vote});
-  }
-
-  db_->saveLastBlockCertVote(vote);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp:526]: If you do not use it anywhere we should remove this function too

File: libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp [hunk 11]
```diff
@@ -480,102 +403,34 @@ uint64_t VoteManager::roundDeterminedFromVotes(size_t two_t_plus_one) {
...
-    LOG(log_er_) << e.what();
-    LOG(log_er_) << "DPOS period " << dpos_period << ", voter " << voter_account_addr << " votes count "
-                 << voter_dpos_votes_count << ", total votes count " << dpos_total_votes_count
-                 << ", pbft sortition threshold " << pbft_sortition_threshold;
-
-    return false;
-  }
-
-  return true;
+bool VoteManager::isInRewardsVotes(const vote_hash_t& vote_hash) {
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp:431]: const ?

File: libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp [hunk 12]
```diff
@@ -152,70 +136,35 @@ class VoteManager {
...
-  bool voteInVerifiedMap(std::shared_ptr<Vote> const& vote);
+  bool voteInVerifiedMap(std::shared_ptr<Vote> const& vote) const;
+
+  /**
+   * @brief Inserts unique vote
+   * @param vote
+   * @return true if vote was successfully inserted(it was unique) or this specific vote was already inserted, otherwise
+   * false
+   */
+  bool insertUniqueVote(const std::shared_ptr<Vote>& vote);
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp:161]: Why `insertUniqueVote` is separate function from `addVerifiedVote` ? is there a case where would you call just one? Because what I can see is that someone will call just one of those and break whole functionality

File: libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp [hunk 13]
```diff
@@ -2,17 +2,147 @@
...
+      vote_mgr_(std::move(vote_mgr)) {}
+
+std::pair<bool, std::string> ExtVotesPacketHandler::validateStandardVote(const std::shared_ptr<Vote> &vote) const {
+  const uint64_t current_pbft_period = pbft_chain_->getPbftChainSize();
+  const auto current_pbft_round = pbft_mgr_->getPbftRound();
+
+  // Old vote or vote from too far in the future, can be dropped
+  // TODO[1880]: should be vote->getPeriod() <= current_pbft_period - if <=, some tests are failing due to missing
+  // reward votes -> whole rewards votes gossiping need to be checked...
+  if (vote->getPeriod() < current_pbft_period || vote->getPeriod() > current_pbft_period + kDposDelay_) {
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp:28]: `kDposDelay_` should be without _

File: libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp [hunk 14]
```diff
@@ -47,7 +176,7 @@ void ExtVotesPacketHandler::sendPbftVotes(const dev::p2p::NodeID &peer_id, std::
     dev::RLPStream s(count);
     for (auto i = index; i < index + count; i++) {
       const auto &v = votes[i];
-      s.appendRaw(v->rlp(true));
+      s.appendRaw(v->rlp(true, true));
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp:179]: why are you include weight?

**Review by MatusKysel — APPROVED** [review]

File: tests/pbft_manager_test.cpp [hunk 15]
```diff
@@ -46,22 +46,6 @@ void check_2tPlus1_validVotingPlayers_activePlayers_threshold(size_t committee_s
...
-      trxs_count++;
-    }
-    EXPECT_HAPPENS({120s, 1s}, [&](auto &ctx) {
-      for (auto &node : nodes) {
-        if (ctx.fail_if(node->getDB()->getNumTransactionExecuted() != trxs_count)) {
-          return;
-        }
-      }
-    });
-  }
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply tests/pbft_manager_test.cpp:64]: - main purpose of function `check_2tPlus1_validVotingPlayers_activePlayers_threshold` is to test dpos transactions ? Either something is terribly wrong with the function name or the main purpose is something totally different
- if we wanna be sure dpos transactions works, there should be separate test for it

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 1 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1150]: https://github.com/Taraxa-project/taraxa-node/issues/1896

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 2 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1193]: will be solved by https://github.com/Taraxa-project/taraxa-node/issues/1896

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 3 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1212]: will be solved by https://github.com/Taraxa-project/taraxa-node/issues/1896

File: tests/pbft_manager_test.cpp (hunk 7 above)
Comment (JakubFornadel) [reply tests/pbft_manager_test.cpp:417]: https://github.com/Taraxa-project/taraxa-node/pull/1900#discussion_r939946599

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 5 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1631]: not anymore. Purpose of this function: @brief Move all verified votes to the unverified map. Since PBFT chain has pushed new blocks, that will affect DPOS. I deleted this function

File: libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp (hunk 10 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp:526]: yeah well it was used just here, but I have no idea why... Any possible clue why we would save block cert vote when adding reward vote ?

File: libraries/core_libs/network/src/network.cpp (hunk 6 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/network.cpp:16]: because in multiple packet handlers I am using "delegation_delay" to know if it even makes sense to process e.g. votes and validate them against dpos contract

File: libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp (hunk 14 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/tarcap/packets_handlers/common/ext_votes_packet_handler.cpp:179]: because the node did not work anymore due to some assert related to empty weight, but I did not spend too much tie thinking about other possible fixes so I can revisit this one

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 2 above)
Comment (MatusKysel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1193]: So please add TODO as we agreed for all things that will be solved ..

File: libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp (hunk 12 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp:161]: the reason is that `insertUniqueVote` is called also when adding reward votes and next votes in NextVotesManager. The other solution is to separate this unique votes functionality into separate class and call it inside `addVerifiedVote`, `addRewardVote` and `updateWithSyncedVotes`

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 2 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:1193]: sure, will do

File: tests/pbft_manager_test.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (MatusKysel) [reply tests/pbft_manager_test.cpp:64]: sorry, I meant that idea of that test was to increase number of eligible nodes in runtime not statistically on start

File: libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp (hunk 10 above)
Comment (MatusKysel) [reply libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp:526]: I guess it was because it could be different cert vote we did not have saved before ?

File: libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp (hunk 12 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/include/vote_manager/vote_manager.hpp:161]: https://github.com/Taraxa-project/taraxa-node/issues/1935

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 16]
```diff
@@ -346,39 +348,46 @@ void PbftManager::resetStep_() {
...
 
 bool PbftManager::resetRound_() {
-  // Jump to round
-  auto consensus_pbft_round = vote_mgr_->roundDeterminedFromVotes(TWO_T_PLUS_ONE);
+  auto determined_round_and_period = vote_mgr_->determineRoundAndPeriodFromVotes(TWO_T_PLUS_ONE);
+  if (!determined_round_and_period.has_value()) {
+    return false;
+  }
+
+  auto [determined_round, determined_previous_round_period] = determined_round_and_period.value();
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:356]: if you do check for optional value, do not use `.value()` but * as it has perf improvements

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 17]
```diff
@@ -509,39 +524,40 @@ void PbftManager::initialState() {
   pbft_step_last_requested_sync_ = 0;
 
   auto own_starting_value = db_->getPbftMgrVotedValue(PbftMgrVotedValue::OwnStartingValueInRound);
-  if (own_starting_value) {
+  if (own_starting_value.has_value()) {
     // From DB
-    own_starting_value_for_round_ = *own_starting_value;
+    own_starting_value_for_round_ = own_starting_value.value();
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:529]: again, this is wrong. if there is check for the value, never use `.value()`

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp [hunk 18]
```diff
@@ -666,84 +682,100 @@ bool PbftManager::stateOperations_() {
...
-    }
-    try {
-      v->validate(dposEligibleVoteCount_(v->getVoterAddr()), getDposTotalVotesCount(), getThreshold(v->getType()), *pk);
-    } catch (const std::logic_error &e) {
-      LOG(log_wr_) << e.what();
-      return false;
+  auto [round, previous_round_period] = getPbftRoundAndPeriod();
+  LOG(log_dg_) << "PBFT current round(r): " << round << ", r-1 period: " << previous_round_period << ", step " << step_;
+
+  static bool is_new_round = false;
```
Comment (MatusKysel) [inline libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:688]: why static?

File: libraries/core_libs/network/src/network.cpp [hunk 19]
```diff
@@ -13,32 +13,33 @@
...
                  std::filesystem::path const &network_file_path, dev::KeyPair const &key, std::shared_ptr<DbStorage> db,
                  std::shared_ptr<PbftManager> pbft_mgr, std::shared_ptr<PbftChain> pbft_chain,
                  std::shared_ptr<VoteManager> vote_mgr, std::shared_ptr<NextVotesManager> next_votes_mgr,
                  std::shared_ptr<DagManager> dag_mgr, std::shared_ptr<TransactionManager> trx_mgr)
-    : kConf(config), tp_(config.network.network_num_threads, false) {
+    : tp_(config.network.network_num_threads, false) {
   auto const &node_addr = key.address();
   LOG_OBJECTS_CREATE("NETWORK");
-  LOG(log_nf_) << "Read Network Config: " << std::endl << kConf.network << std::endl;
+  LOG(log_nf_) << "Read Network Config: " << std::endl << config.network << std::endl;
```
Comment (MatusKysel) [inline libraries/core_libs/network/src/network.cpp:25]: why did you renamed constant member?

**Review by mfrankovi — APPROVED** [review]

File: libraries/core_libs/network/src/network.cpp (hunk 19 above)
Comment (JakubFornadel) [reply libraries/core_libs/network/src/network.cpp:25]: it is not a member anymore, it is ctor argument

File: libraries/core_libs/consensus/src/pbft/pbft_manager.cpp (hunk 18 above)
Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:688]: because I need it to hold the value throughout multiple calls. I could either have class member or static variable - as this variable is needed only in this single function I used static rather than class member

Comment (MatusKysel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp:688]: ok, but I do not think it is good idea, as now you have some state, that it is shared between all instances of pbft manager and it should be bind to specific instance not to all. Doesn't it influence tests?

Comment (JakubFornadel) [reply libraries/core_libs/consensus/src/pbft/pbft_manager.cpp]: true, fixed
